### PR TITLE
Simplify DbProviderFactory feature detection code

### PR DESCRIPTION
**Pull request by bgrainger**
See comment: https://github.com/dotnet/corefx/commit/60d65afa0160a0b27e8df6b8f450ead8ab7f4214#r33736759

### CODE REVIEW

File: src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs [hunk 1]
```diff
@@ -19,15 +19,9 @@ public virtual bool CanCreateDataAdapter
...
-                    if (adapter == null)
+                    using (var adapter = CreateDataAdapter())
                     {
-                        _canCreateDataAdapter = false;
-                    }
-                    else
-                    {
-                        _canCreateDataAdapter = true;
-                        adapter.Dispose();
+                        _canCreateDataAdapter = adapter is object;
```
Comment (stephentoub) [inline src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs]: We've generally stayed away from this form of not-null check, as it's not obvious to most readers what it's actually verifying.  Please change it to `adapter != null`.

File: src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs [hunk 2]
```diff
@@ -41,15 +35,9 @@ public virtual bool CanCreateCommandBuilder
...
-                    if (builder == null)
-                    {
-                        _canCreateCommandBuilder = false;
-                    }
-                    else
+                    using (var builder = CreateCommandBuilder())
                     {
-                        _canCreateCommandBuilder = true;
-                        builder.Dispose();
+                        _canCreateCommandBuilder = builder is object;
```
Comment (stephentoub) [inline src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs]: Ditto

File: src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs [hunk 3]
```diff
@@ -41,15 +35,9 @@ public virtual bool CanCreateCommandBuilder
             {
                 if (!_canCreateCommandBuilder.HasValue)
                 {
-                    var builder = CreateCommandBuilder();
-                    if (builder == null)
-                    {
-                        _canCreateCommandBuilder = false;
-                    }
-                    else
+                    using (var builder = CreateCommandBuilder())
```
Comment (stephentoub) [inline src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs]: Nit: var => whatever the type actually is

File: src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs [hunk 4]
```diff
@@ -19,15 +19,9 @@ public virtual bool CanCreateDataAdapter
             {
                 if (!_canCreateDataAdapter.HasValue)
                 {
-                    var adapter = CreateDataAdapter();
-                    if (adapter == null)
+                    using (var adapter = CreateDataAdapter())
```
Comment (stephentoub) [inline src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs]: Nit: var => whatever the type actually is

File: src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs (hunk 1 above)
Comment (bgrainger) [reply src/System.Data.Common/src/System/Data/Common/DbProviderFactory.cs]: I did wonder about that; will change. Thanks.

**Pull request merged**